src/orange/OrangeTrackView.hh

```diff
@@ -491,19 +551,95 @@ CELER_FUNCTION void OrangeTrackView::set_dir(Real3 const& newdir)
         {
             // The boundary crossing direction has changed! Reverse our plans
             // to change the logical state and move to a new volume.
-            states_.boundary[thread_]
-                = flip_boundary(states_.boundary[thread_]);
+            lsa.boundary() = flip_boundary(lsa.boundary());
         }
     }
 
     // Complete direction setting
-    states_.dir[thread_] = newdir;
+    lsa.dir() = newdir;
 
     this->clear_next_step();
 }
 
 //---------------------------------------------------------------------------//
 // PRIVATE MEMBER FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Iterate over all levels to find the next step.
+ */
+CELER_FUNCTION void OrangeTrackView::find_next_step_impl(real_type max_step)
+{
+    // The univese the particle is currently within
+
+    auto current_level_lsa = this->make_lsa();
+    auto const& current_uid = current_level_lsa.universe();
+
+    // The uid we are iteratively checking for nearest intersection
+    UniverseId check_uid;
+
+    // The next uid we will check
+    UniverseId next_check_uid = top_universe_id();
+
+    auto min_step = max_step;
+    celeritas::detail::OnSurface min_surface_local;
+    UniverseId min_uid;
+
+    size_type level = 0;
+
+    do
+    {
+        check_uid = next_check_uid;
+        auto tracker = this->make_tracker(check_uid);
+        auto isect = tracker.intersect(this->make_local_state(LevelId{level}),
+                                       max_step);
+
+        if (isect.distance < min_step)
+        {
+            min_step = isect.distance;
+            min_surface_local = isect.surface;
+            min_uid = check_uid;
+        }
+
+        auto lsa = this->make_lsa(LevelId{level});
+        next_check_uid = params_.volume_records[lsa.vol()].daughter;
+        ++level;
+    } while (check_uid != current_uid);
+
+    next_step_ = min_step;
+
+    // convert local to global surface
```

Comment:
Comments should generally be like sentences (usually omitting the `.` at the end though): capitalize the first letter.